Optimize snippet()

topecongiro opened the pull request:
## Motivation
rustfmt makes heavy use of `snippet()` to fetch a piece of snippet from the original file. It returns `String`, even though we do not modify the returned value. It is preferable to return `&str` to avoid allocations on heap.

## Pros

We get about 15% speed up against a large code base. The following is a quick experiment result that runs `cargo fmt --all` against rustc:

### env

- OS: Linux 4.13.16
- rustc: `rustc 1.24.0-nightly (1956d5535 2017-12-03)`
- CPU: `Intel(R) Core(TM) i3-4150 CPU @ 3.50GHz`
- RAM: 8GB

### Result

| version | time [s] (an average of 10 attempts) |
| ----------|-------|
| master (0432d4db35cee640ca96460f4ebc4baad3891d6d) | 9.77 |
| PR (a5b5ef078bd835f68a1e2b8eb5f63d0797e4e0e8) | 8.47 |

## Cons

This PR introduces `unsafe` code in rustfmt (specifically in `impl SnippetProducer`):

- transmute `&Option<Rc<String>>` into `*const Rc<String>`.
- dereference `*const`.

nrc commented on src/visitor.rs:
```diff
@@ -47,6 +49,32 @@ fn is_extern_crate(item: &ast::Item) -> bool {
+
+    pub fn from_codemap(codemap: &CodeMap, span: Span) -> Self {
+        let filemap = codemap.lookup_char_pos(span.lo()).file;
+        let big_snippet = unsafe { mem::transmute(&filemap.src) };
```
Rather than do this transmute, could we unwrap the option (or return Option<SnippetProvider>) and keep an Rc rather than `*const Rc<String>`? This looks pretty bad because we never increment the reference count on the Rc pointer, plus the transmute is relying on implementation details of the Rust compiler.

nrc commented on src/visitor.rs:
```diff
@@ -47,6 +49,32 @@ fn is_extern_crate(item: &ast::Item) -> bool {
+}
+
+impl SnippetProvider {
+    pub fn span_to_snippet(&self, span: Span) -> Option<&str> {
```
Could we use an API such as `with_snippet_for_span` which takes a closure `Fn(&str) -> T` thus making a safe interface here?

topecongiro commented:
Thank you for the reviews! The last two commits were not supposed to be in this PR, should have been put in another PR. My apologies for the confusion.

I removed the last two commits and updated the PR to get rid of `unsafe` code.

sinkuu commented on src/visitor.rs:
```diff
@@ -47,6 +48,30 @@ fn is_extern_crate(item: &ast::Item) -> bool {
+/// Creates a string slice corresponding to the specified span.
+pub struct SnippetProvider<'a> {
+    /// A pointer to the content of the file we are formatting.
+    big_snippet: &'a Rc<String>,
```
Any reason for not cloning `Rc` (and `big_snippet: Rc<String>`)? It is just an increment, and you can get rid of a layer of indirection ("`&` -> `Rc` -> `String`" to "`Rc` -> `String`"), which might be good for performance.

sinkuu replied:
This seems to work and simpler:

```diff
diff --git a/src/lib.rs b/src/lib.rs
index f6902db..87fcc0b 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -322,7 +322,7 @@ where
             .lookup_char_pos(module.inner.lo())
             .file;
         let big_snippet = filemap.src.as_ref().unwrap();
-        let snippet_provider = SnippetProvider::new(filemap.start_pos, big_snippet);
+        let snippet_provider = SnippetProvider::new(filemap.start_pos, &big_snippet);
         let mut visitor = FmtVisitor::from_codemap(parse_session, config, &snippet_provider);
         // Format inner attributes if available.
         if !krate.attrs.is_empty() && path == main_file {
diff --git a/src/visitor.rs b/src/visitor.rs
index f2b4750..d757037 100644
--- a/src/visitor.rs
+++ b/src/visitor.rs
@@ -51,7 +51,7 @@ fn is_extern_crate(item: &ast::Item) -> bool {
 /// Creates a string slice corresponding to the specified span.
 pub struct SnippetProvider<'a> {
     /// A pointer to the content of the file we are formatting.
-    big_snippet: &'a Rc<String>,
+    big_snippet: &'a str,
     /// A position of the start of `big_snippet`, used as an offset.
     start_pos: usize,
 }
@@ -63,7 +63,7 @@ impl<'b, 'a: 'b> SnippetProvider<'a> {
         Some(&self.big_snippet[start_index..end_index])
     }
 
-    pub fn new(start_pos: BytePos, big_snippet: &'a Rc<String>) -> Self {
+    pub fn new(start_pos: BytePos, big_snippet: &'a str) -> Self {
         let start_pos = start_pos.to_usize();
         SnippetProvider {
             big_snippet,
```

topecongiro commented:
@sinkuu Thanks for your review!

Updated with @sinkuu's review reflected.

nrc commented on src/visitor.rs:
```diff
@@ -47,6 +47,30 @@ fn is_extern_crate(item: &ast::Item) -> bool {
+}
+
+impl<'b, 'a: 'b> SnippetProvider<'a> {
+    pub fn span_to_snippet(&'b self, span: Span) -> Option<&'a str> {
```
Do you need `'b'` here? It looks like you shouldn't

nrc commented on src/patterns.rs:
```diff
@@ -122,7 +122,7 @@ impl Rewrite for Pat {
             }
             // FIXME(#819) format pattern macros.
-            PatKind::Mac(..) => Some(context.snippet(self.span)),
+            PatKind::Mac(..) => Some(context.snippet(self.span).into()),
```
Are all these `into`s converting `&str` to `String`? If so, could you use `to_owned` instead please?

topecongiro commented:
@nrc Thanks again, updated with the review reflected.

nrc commented:
Thanks for all the changes, and thanks for the PR - it's a great perf improvement!
